src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Minmod.hpp

```diff
@@ -0,0 +1,309 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/Element.hpp"  // IWYU pragma: keep
+#include "ErrorHandling/Assert.hpp"
+#include "NumericalAlgorithms/LinearOperators/MeanValue.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+class DataVector;
+template <size_t VolumeDim>
+class Direction;
+template <size_t>
+class Mesh;
+
+namespace PUP {
+class er;
+}  // namespace PUP
+
+namespace SlopeLimiters {
+template <size_t VolumeDim, typename TagsToLimit>
+class Minmod;
+}  // namespace SlopeLimiters
+/// \endcond
+
+namespace SlopeLimiters {
+/// \ingroup SlopeLimitersGroup
+/// \brief Possible types of the minmod slope limiter.
+///
+/// \see The documentation of `Minmod`, where these types are discussed.
+enum class MinmodType { LambdaPi1, LambdaPiN, Muscl };
+}  // namespace SlopeLimiters
+
+namespace Minmod_detail {
+// Encodes the return status of the minmod_tvbm function.
+struct MinmodResult {
+  const double value;
+  const bool activated;
+};
+
+// The TVBM-corrected minmod function, see e.g. Cockburn reference Eq. 2.26.
+MinmodResult minmod_tvbm(double a, double b, double c,
+                         double tvbm_scale) noexcept;
+
+// Implements the minmod limiter for one Tensor<DataVector>.
+//
+// The interface is designed to erase the tensor structure information, because
+// this way the implementation can be moved out of the header file. This is
+// achieved by receiving Tensor<DataVector>::iterators into the tensor to limit,
+// and Tensor<double>::iterators into the neighbor tensors.
+//
+// Note: because the interface erases the tensor structure information, we can
+// no longer rely on the compiler to enforce that the local and neighbor tensors
+// share the same Structure.
+template <size_t VolumeDim>
+bool limit_one_tensor(
+    DataVector* tensor_begin, DataVector* tensor_end,
```

Comment:
@nilsdeppe These probably shouldn't stay as raw pointers. I don't really want to stick iterators into `not_null`s. I could make `limit_one_tensor` accept `Scalar<DataVector>::iterator tensor_begin` etc.... the wrong Tensor type, but still the correct iterator type (but risky if we ever change Tensor so that different Structures have different iterators). Thoughts?

Reply:
Well, the "most correct" is probably implementing a TensorIterator, which is not something we should spend time on right now. I'm fine with these as raw or not_null pointers since it's inside an implementation and not user-facing.

[nitpick] can the pointers be const?

Reply:
I agree; let's hold off on implementing a type-erasing Tensor iterator class until we know this would be useful in more than one place.

Re [nitpick] The data pointed to cannot be const, because the limiting occurs in-place. The pointers themselves are const in the cpp, but clang-tidy doesn't want them const in the header.

Reply:
derp, thanks :)

Reply:
Went with `not_null`s for consistency with rest of codebase.